### PR TITLE
Feature/shikomori track

### PR DESCRIPTION
**Pull request by Pavkazzz**
Add russian track - https://shikimori.org/

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/data/track/TrackManager.kt [hunk 1]
```diff
@@ -4,13 +4,15 @@ import android.content.Context
...
 import eu.kanade.tachiyomi.data.track.myanimelist.Myanimelist
+import eu.kanade.tachiyomi.data.track.shikomori.Shikomori
 
 class TrackManager(private val context: Context) {
 
     companion object {
         const val MYANIMELIST = 1
         const val ANILIST = 2
         const val KITSU = 3
+        const val SHIKOORI = 4
```
Comment (Taumer) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/TrackManager.kt]: Missed `M`

File: app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import uy.kohesive.injekt.injectLazy
+
+class ShikomoriApi(private val client: OkHttpClient, interceptor: ShikomoriInterceptor) {
+
+    private val gson: Gson by injectLazy()
+    private val parser = JsonParser()
+    private val jsonime = MediaType.parse("application/json; charset=utf-8")
+    private val authClient = client.newBuilder().addInterceptor(interceptor).build()
+
+    fun addLibanga(track: Track, user_id: String): Observable<Track> {
```
Comment (Taumer) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt]: `M`?

File: app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                .post(body)
+                .build()
+        return authClient.newCall(request)
+                .asObservableSuccess()
+                .map {
+                    track
+                }
+    }
+
+    fun updateLibanga(track: Track, user_id: String): Observable<Track> = addLibanga(track, user_id)
```
Comment (Taumer) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt]: `M`

File: app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            media_id = obj["id"].asInt
+            title = ""
+            last_chapter_read = obj["chapters"].asInt
+            total_chapters = obj["chapters"].asInt
+            score = (obj["score"].asInt).toFloat()
+            status = toTrackStatus(obj["status"].asString)
+        }
+    }
+
+    fun findLibanga(track: Track, user_id: String): Observable<Track?> {
```
Comment (Taumer) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt]: `M`

File: app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        private const val clientId = "1aaf4cf232372708e98b5abc813d795b539c5a916dbbfe9ac61bf02a360832cc"
+        private const val clientSecret = "229942c742dd4cde803125d17d64501d91c0b12e14cb1e5120184d77d67024c0"
+
+        private const val baseUrl = "https://shikimori.org"
+        private const val apiUrl = "https://shikimori.org/api"
+        private const val oauthUrl = "https://shikimori.org/oauth/token"
+        private const val loginUrl = "https://shikimori.org/oauth/authorize"
+
+        private const val redirectUrl = "tachiyomi://shikimori-auth"
+        private const val baseangaUrl = "$apiUrl/mangas"
```
Comment (Taumer) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt]: `M`

Comment (Pavkazzz) [reply app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt]: 🤦‍♂️

File: app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            score = (obj["score"].asInt).toFloat()
+            status = toTrackStatus(obj["status"].asString)
+        }
+    }
+
+    fun findLibanga(track: Track, user_id: String): Observable<Track?> {
+        val url = Uri.parse("$apiUrl/v2/user_rates").buildUpon()
+                .appendQueryParameter("user_id", user_id)
+                .appendQueryParameter("target_id", track.media_id.toString())
+                .appendQueryParameter("target_type", "anga")
```
Comment (Taumer) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/ShikomoriApi.kt]: `M`

**Comment by Pavkazzz** [comment]
@inorichi Test this already, can you merge this?

File: app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/OAuth.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+data class OAuth(
+        val access_token: String,
+        val token_type: String,
+        val created_at: Long,
+        val expires_in: Long,
+        val refresh_token: String?) {
+
+    // Access token lives 1 day
+    fun isExpired() = (System.currentTimeMillis() / 1000) > (created_at + expires_in - 3600)
```
Comment (CarlosEsco) [inline app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/OAuth.kt]: Does this mean user has to login in daily? Man that's a hassle

Comment (Pavkazzz) [reply app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/OAuth.kt]: No, it's about refresh it, each 24 hours

Comment (Pavkazzz) [reply app/src/main/java/eu/kanade/tachiyomi/data/track/shikomori/OAuth.kt]: It's very strange politics, in my opinion :)

**Pull request merged**